src/software/simulated_tests/validation/non_terminating_function_validator.cpp

```diff
@@ -30,6 +32,11 @@ void NonTerminatingFunctionValidator::executeAndCheckForFailures()
     // Run the coroutine. This will call the bound executeAndCheckForFailuresWrapper
     // function
     validation_sequence();
+    std::string error_msg = validation_sequence.get();
+    if (error_msg != "")
+    {
+        ADD_FAILURE() << error_msg;
```

Comment:
My only real nit is that `ADD_FAILURE` is called at different levels for terminating vs non-terminating validation functions. For terminating validation functions, `ADD_FAILURE` is called at the test-fixture level if there are errors to report, while the non-terminating validation functions call it here inside the validator object.

I understand why it was done this way because non-terminating conditions need to constantly report if they have passed or not (since they should always pass).

It may be more clear to also give the non-terminating validation funcs a "latest error message" function that can be used by the test fixture to report the errors instead. This is mostly for consistency because it initially confused me a bit and I had to go back and double check where the failures for each type of validation function was added.

Reply:
I made it return an optional string